pulse/broker/src/main/java/gov/ca/emsa/pulse/broker/dao/impl/PatientNameDAOImpl.java

```diff
@@ -0,0 +1,154 @@
+import gov.ca.emsa.pulse.broker.entity.AddressEntity;
+import gov.ca.emsa.pulse.broker.entity.AlternateCareFacilityEntity;
+import gov.ca.emsa.pulse.broker.entity.GivenNameEntity;
+import gov.ca.emsa.pulse.broker.entity.PatientEntity;
+import gov.ca.emsa.pulse.broker.entity.PatientNameEntity;
+import gov.ca.emsa.pulse.broker.entity.PatientOrganizationMapEntity;
+
+@Repository
+public class PatientNameDAOImpl extends BaseDAOImpl implements PatientNameDAO {
+	private static final Logger logger = LogManager.getLogger(PatientNameDAOImpl.class);
+	@Autowired GivenNameDAO givenDAO;
+
+	@Override
+	public PatientNameDTO create(PatientNameDTO dto) {
+		PatientNameEntity patient = new PatientNameEntity();
+		patient.setFamilyName(dto.getFamilyName());
+		ArrayList<GivenNameDTO> givensDTO = new ArrayList<GivenNameDTO>();
+		if(dto.getGivenName() != null){
+			for(GivenNameDTO givenDto : dto.getGivenName()){
+				if(givenDto.getId() == null){
+					GivenNameDTO givenName = givenDAO.create(givenDto);
+					givensDTO.add(givenName);
+				}
+			}
+			dto.setGivenName(givensDTO);
+			//given name entities should exist now
+			ArrayList<GivenNameEntity> givensEntity = new ArrayList<GivenNameEntity>();
+			for(GivenNameDTO given: dto.getGivenName()){
+				GivenNameEntity name = entityManager.find(GivenNameEntity.class, given.getId());
+				givensEntity.add(name);
+			}
+			patient.setGivenNames(givensEntity);
+		}
+		patient.setSuffix(dto.getSuffix());
+		patient.setPrefix(dto.getPrefix());
+		patient.setProfSuffix(dto.getProfSuffix());
+		patient.setNameTypeCode(dto.getNameTypeCode());
+		patient.setNameTypeCodeDescription(dto.getNameTypeCodeDescription());
+		patient.setNameRepresentationCode(dto.getNameRepresentationCode());
+		patient.setNameRepresentationCodeDescription(dto.getNameRepresentationCodeDescription());
+		patient.setNameAssemblyOrderCode(dto.getNameAssemblyOrderCode());
+		patient.setNameAssemblyOrderCodeDescription(dto.getNameAssemblyOrderCodeDescription());
+		patient.setEffectiveDate(dto.getEffectiveDate());
+		patient.setExpirationDate(dto.getExpirationDate());
+
+		entityManager.persist(patient);
+		entityManager.flush();
+		return new PatientNameDTO(patient);
+	}
+
+	@Override
+	public PatientNameDTO update(PatientNameDTO dto) {
+		PatientNameEntity patient = this.getEntityById(dto.getId());
+		patient.setFamilyName(dto.getFamilyName());
+		ArrayList<GivenNameEntity> givens = new ArrayList<GivenNameEntity>();
+		for(GivenNameDTO givenDto : dto.getGivenName()){
+			GivenNameEntity givenName = new GivenNameEntity();
+			givenName.setGivenName(givenDto.getGivenName());
+			givenName.setId(givenDto.getId());
+			givenName.setPatientNameId(givenDto.getPatientNameId());
+			givens.add(givenName);
+		}
+		patient.setGivenNames(givens);
+		if(dto.getSuffix() != null)
+			patient.setSuffix(dto.getSuffix());
+		if(dto.getPrefix() != null)
+			patient.setPrefix(dto.getPrefix());
+		if(dto.getNameTypeCode() != null)
```

Comment:
There are a few types of values for a name that are lookup values: Name Type Code, Name Representation, Name Assembly Order, maybe one other. Those should not be stored in the PatientName table directly, they should a foreign key relationship to three (or more) other, separate tables